backend/src/main/java/gov/cdc/usds/simplereport/utils/BulkUploadResultsToHL7.java

```
@@ -0,0 +1,524 @@
+
+    return String.join(
+        "|",
+        fields.fieldName(),
+        ENCODING_CHARS,
+        sendingApplication,
+        sendingFacility,
+        receivingApplication,
+        receivingFacility,
+        fields.datetime());
+  }
+
+  private String generateTrailingSegment(String fieldName, int itemCount) {
+    return String.join("|", fieldName, String.valueOf(itemCount));
+  }
+
+  public HL7BatchMessage convertToHL7BatchMessage(InputStream csvStream) {
+    String batchMessage = "";
+    int batchMessageCount = 0;
+    HashMap<String, Integer> diseasesReported = new HashMap<>();
+    String hl7Date = formatToHL7DateTime(dateGenerator.newDate());
+    var futureTestEvents = new ArrayList<CompletableFuture<String>>();
+    final MappingIterator<Map<String, String>> valueIterator = getIteratorForCsv(csvStream);
+
+    String sendingFacilityNamespaceId = null;
+    String sendingFacilityUniversalId = null;
+
+    while (valueIterator.hasNext()) {
+      final Map<String, String> row = getNextRow(valueIterator);
+      TestResultRow fileRow = new TestResultRow(row);
+
+      if (sendingFacilityNamespaceId == null) {
+        sendingFacilityNamespaceId = fileRow.getTestingLabName().getValue();
+      }
+
+      if (sendingFacilityUniversalId == null) {
+        sendingFacilityUniversalId = fileRow.getTestingLabClia().getValue();
+      }
+
+      Optional<String> disease =
+          getDiseaseFromDeviceSpecs(
+              fileRow.getEquipmentModelName().getValue(),
+              fileRow.getTestPerformedCode().getValue());
+
+      if (disease.isPresent()) {
+        diseasesReported.put(disease.get(), diseasesReported.getOrDefault(disease.get(), 0) + 1);
+      }
+
+      var future =
+          CompletableFuture.supplyAsync(
+              () -> {
+                try {
+                  return convertRowToHL7(fileRow);
+                } catch (HL7Exception e) {
+                  // return empty string for message
```

Should we throw an error here?

---

I think so after dwelling on it more. This would mean that if any 1 row fails to be converted to HL7, the whole upload should fail. Which I think makes sense if we care about data integrity.

---

That makes sense.